File: core/Sqrtx/Sqrtx.py
def mySqrt(x: int) -> int:
    if (x // 2) ** 2 < x:
        return x // 2 if x != 1 else 1

    left, right = 0, x // 2
    while left ** 2 <= x <= right ** 2:
        mid = (left + right) // 2
        if mid ** 2 == x:
            return mid

        if mid ** 2 > x:
            right = mid - 1
        else:
            left = mid + 1

    if left ** 2 > x:
        return left - 1 if left ** 2 > x else left
    if right ** 2 < x:
        return right if right ** 2 < x else right - 1

if __name__ == '__main__':
    print(mySqrt(8192))

File: core/Sqrtx/Sqrtx.java
package com.dbc;

public class Sqrtx {
    public int mySqrt(int x) {
        if (x == 0) return 0;
        if (x == 1) return 1;
        int left = 0, right = x / 2;
        while (Math.pow(left, 2) <= x && x <= Math.pow(right, 2)) {
            int mid = (left + right) / 2;
            if (Math.pow(mid, 2) == x) return mid;
            if (Math.pow(mid, 2) > x) right = mid - 1;
            else left = mid + 1;
        }
        if (Math.pow(left, 2) > x) return left - 1;
        else return right;
    }
}
